plugin/evm/vm_warp_test.go

```diff
@@ -408,74 +410,142 @@ func TestReceiveWarpMessage(t *testing.T) {
 	}
 	genesisJSON, err := genesis.MarshalJSON()
 	require.NoError(err)
-	issuer, vm, _, _ := GenesisVM(t, true, string(genesisJSON), "", "")
+
+	// disable warp so we can re-enable it with RequirePrimaryNetworkSigners
+	disableTime := upgrade.InitiallyActiveTime.Add(10 * time.Second)
+	disableConfig := warp.NewDisableConfig(utils.TimeToNewUint64(disableTime))
+
+	// re-enable warp with RequirePrimaryNetworkSigners
+	reEnableTime := disableTime.Add(10 * time.Second)
+	reEnableConfig := warp.NewDefaultConfig(utils.TimeToNewUint64(reEnableTime)).WithRequirePrimaryNetworkSigners(true)
+
+	upgradeConfig := params.UpgradeConfig{
+		PrecompileUpgrades: []params.PrecompileUpgrade{
+			{Config: disableConfig},
+			{Config: reEnableConfig},
+		},
+	}
+	upgradeBytes, err := json.Marshal(upgradeConfig)
+	require.NoError(err)
+
+	issuer, vm, _, _ := GenesisVM(t, true, string(genesisJSON), "", string(upgradeBytes))
 
 	defer func() {
 		require.NoError(vm.Shutdown(context.Background()))
 	}()
 
-	acceptedLogsChan := make(chan []*types.Log, 10)
-	logsSub := vm.eth.APIBackend.SubscribeAcceptedLogsEvent(acceptedLogsChan)
-	defer logsSub.Unsubscribe()
+	// Subnet messages should use subnet signers
+	testReceiveWarpMessage(t, issuer, vm, false, false, upgrade.InitiallyActiveTime)
 
-	payloadData := avagoUtils.RandomBytes(100)
+	// At genesis, primary network messages should use subnet signers
+	blockTime := upgrade.InitiallyActiveTime.Add(2 * time.Second) // for fees
+	require.True(blockTime.Before(disableTime))
+	testReceiveWarpMessage(t, issuer, vm, true, false, blockTime)
 
+	// After re-enabled, primary network messages should use primary signers
+	testReceiveWarpMessage(t, issuer, vm, true, true, reEnableTime)
+}
+
+func testReceiveWarpMessage(
+	t *testing.T, issuer chan commonEng.Message, vm *VM,
+	fromPrimary, usePrimarySigners bool,
```

Comment:
These booleans make it difficult to understand the call sites (made me review out of order because I didn't know what was going on). One fix is to introduce enums:

```go
type warmMsgFrom int

const (
    fromSubnet warpMsgFrom = iota
    fromPrimary
)

type useWarpMsgSigners int

const (
    subnetSigners warpMsgSigners = iota
    primarySigners
)
```

Then the call site is clear:

```go
testReceiveWarpMessage(t, issuer, vm, fromSubnet, subnetSigners, upgrade.InitiallyActiveTime)
```

Also see [Uber style guide](https://github.com/uber-go/guide/blob/master/style.md#avoid-naked-parameters).